### DISCUSSION TITLE
Issues with certain functions after switch from sync to async

### DISCUSSION THREAD

**Discussion by MichaelTatlock** [webdriverio/webdriverio#9310] [Q&A]
I'm in the process of switching my wdio framework from sync to async using the codemod provided and I'm running into a couple issues with certain functions that stopped working. 

The following test is meant to do a search (which it does correctly) and then take the results and compare (where it seems to fail). 

Running:
``` 
WDIO v7.20.7
Nodejs v18.12.1
Java v17.0.5 LTS
```
Below is a snippet of my spec file: 

```
describe('Games Bet Query Tests', () => {
  before(async () => {
    await GamesBetQuery.openBackOffice();
  });

  describe('C9647411 - As an Admin, I can search Games Bet by Game Class', () => {
    it('finds the games bet for the specified Game Class', async () => {
      await GamesBetQuery.search({ gameClass: 'BJ', startedFrom: startDate1, startedTill: startDate2 });
      await assert(GamesBetResult.results.every((bet) => GamesBetResult.getGameClass(bet) === 'BJ'),
        'Expected: All Games Bet found have Game Class BJ but did not');
    });

});
```
It references the following code from page object file: 

```class GamesBetQueryResultPage extends BackOfficePage {
   get results()    { return $$('form > table > tbody > tr').slice(2); }
```
And
```
  /**
   * getGameClass() returns the Game Class for the specified Games Bet
   *
   * @param {Object} bet a WDIO object pointing to the <tr> element for the Games Bet result
   * @returns {String} the Game Class for the specified Games Bet
   */
  async getGameClass(bet) {
    return (await bet.$$('td'))[13].getText();
  }
```
After running my spec file test it reports the following error: 
```
TypeError: GamesBetResult.results.every is not a function
```
This code was working when running sync, so I assume it just has to do with the awaits. I'm fairly new to all of this, so please forgive me if I've missed something and let me know if more information is needed. Thanks for any help you can provide.

**Comment by christian-bromann** [comment]
Change this:

```patch
- assert assert(GamesBetResult.results.every((bet) => GamesBetResult.getGameClass(bet) === 'BJ'),
+ assert((await GamesBetResult.results).every((bet) => GamesBetResult.getGameClass(bet) === 'BJ'),
        'Expected: All Games Bet found have Game Class BJ but did not');
```

You need to know where the promise is and use `await` for that.

**Reply by MichaelTatlock** [reply]
Just to clarify, remove the first await from the assert and place it where you have it? 

If so, I've done such and now I see that the test is failing the expected with the following error: 

```
AssertionError [ERR_ASSERTION]: Expected: All Games Bet found have Game Class BJ but did not
```
Seems like it doesn't find the text.

**Reply by christian-bromann** [reply]
Seems like next step is to debug what `await GamesBetResult.results)` actually is 😉

**Reply by MichaelTatlock** [reply]
I believe the GamesBetResults.results is calling the selector shown in my page object file for the results page: 
```
get results()    { return $$('form > table > tbody > tr').slice(2); }
````
That selector is just an array from a column on my results webpage. If I do a console.log on just GamesBetResults.results after the search it returns a lot of information that doesn't seem right to me: 
```
[0-0] [
[0-0]   Element {
[0-0]     sessionId: '965c431b038b863e39248f2c48a8bf96',
[0-0]     elementId: 'd28bfbde-c541-4ca2-b301-32b967806ad5',
[0-0]     'element-6066-11e4-a52e-4f735466cecf': 'd28bfbde-c541-4ca2-b301-32b967806ad5',
[0-0]     selector: 'form > table > tbody > tr',
[0-0]     parent: Browser {
[0-0]       sessionId: '965c431b038b863e39248f2c48a8bf96',
[0-0]       capabilities: [Object],
[0-0]       addCommand: [Function (anonymous)],
[0-0]       overwriteCommand: [Function (anonymous)],
[0-0]       addLocatorStrategy: [Function (anonymous)],
[0-0]       config: [Object]
[0-0]     },
[0-0]     index: 2,
[0-0]     emit: [Function: bound ],
[0-0]     isReactElement: false,
[0-0]     addCommand: [Function (anonymous)],
[0-0]     overwriteCommand: [Function (anonymous)]
[0-0]   },
[0-0]   Element {
[0-0]     sessionId: '965c431b038b863e39248f2c48a8bf96',
[0-0]     elementId: '665e90e4-2bd7-4cc5-be10-91b37ce53caa',
[0-0]     'element-6066-11e4-a52e-4f735466cecf': '665e90e4-2bd7-4cc5-be10-91b37ce53caa',
[0-0]     selector: 'form > table > tbody > tr',
[0-0]     parent: Browser {
[0-0]       sessionId: '965c431b038b863e39248f2c48a8bf96',
[0-0]       capabilities: [Object],
[0-0]       addCommand: [Function (anonymous)],
[0-0]       overwriteCommand: [Function (anonymous)],
[0-0]       addLocatorStrategy: [Function (anonymous)],
[0-0]       config: [Object]
[0-0]     },
[0-0]     index: 3,
[0-0]     emit: [Function: bound ],
[0-0]     isReactElement: false,
[0-0]     addCommand: [Function (anonymous)],
[0-0]     overwriteCommand: [Function (anonymous)]
[0-0]   }
[0-0] ]
```


From what I can gather, the assert I have should call the results selector and then the getGameClass function and make sure it returns true (i.e. the game class BJ is the only text appearing in the results column). 

Sorry if I'm unsure on certain things as this is existing code that was written by someone else (doesn't work for us anymore) that I've converting to work with async.

**Comment by christian-bromann** [comment]
`$$('form > table > tbody > tr').slice(2)` this doesn't work because `$$` returns a promise. So this needs to be `(await $$('form > table > tbody > tr')).slice(2)`

**Reply by MichaelTatlock** [reply]
Hmm, that's my selector at the top of the page object file, which would explain why it doesn't work now but worked when running sync. How does one fix that as I can't put an await in there.

**Reply by christian-bromann** [reply]
move it into a function or do `$$('form > table > tbody > tr').filter((elem, i) => i < 2)`

**Reply by MichaelTatlock** [reply]
Thanks for the advice. I've tried what you've suggested and unfortunately I have had zero luck. I have narrowed things down, though. 

I've managed to to get it so it will read the array of results on the search, but for some reason the .every function gets called before it has time to get the text of the results for the game class. 

`await assert((await GamesBetResult.results).every((bet) => (GamesBetResult.getGameClass(bet)) === 'BJ'),
        'Expected: All Games Bet found have Game Class BJ but did not');`

In this line of code, it will fetch the results,  but the .every comes back as false due to the getGameClass function being called after the check. Do you have any suggestion on how to get the getGameClass function to be called before the .every true/false is checked? 

Below is the getGameClass function found in the page object file: 
```
 /**
   * getGameClass() returns the Game Class for the specified Games Bet
   *
   * @param {Object} bet a WDIO object pointing to the <tr> element for the Games Bet result
   * @returns {String} the Game Class for the specified Games Bet
   */
  async getGameClass(bet) {
    return (await bet.$$('td'))[13].getText();
  }
```
